### PR TITLE
[Autoscaler/Core] Drain node API

### LINKED ISSUE #19056: [Core] support graceful ray node shutdown
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Currently the only way to gracefully shutdown Ray node is by running `ray stop` on raylet node. However this could be challenging for autoscaler, since it might need to ssh onto that node, which is tricky to make it correct.

As a result, autoscaler kills node ungracefully and propagates node failure message to the driver; which is misleading and hurts overall usability.

Instead, we should probably have a RPC call to gracefully shutdown a node, either through raylet or through GCS, which make autoscaler easy to work with.

### Use case

autoscaler

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR implements "Ray draining API". Note that "drain node" means we will make the requested node invisible to the cluster. 
Drain doesn't mean it kills raylets. Killing raylet could be a implementation detail of draining nodes. 

The PR consists of 3 parts. 
1. **(autoscaler team needs to review)** Implement the DrainNode RPC-based API the autoscaler can consume.
2. **(core team needs to review)** Implement KillRaylet RPC which kills the raylet in the same way as `ray stop`. 
3. **(core team needs to review)** Implement DrainNode API itself. 

### API guarantee
When the RPC is replied from GCS (ack from GCS),
- It is guaranteed that the requested nodes are going to be drained eventually.
- It is guaranteed that GCS has persisted the draining states.
- [RFC] It is **not** guarateed that nodes receive the drain requests from GCS (but it eventually will). 
   - note that this means autoscaler can kill raylets before it receives the drain request from GCS. But since the GCS already persisted the fact that the node was dead, there should be no false negative error messages. 
   - If we guarantee this property, the drain node RPC can be extremely slow if some of nodes have high load. This can make the failure behavior very complex and might require the core to implement `timeout`. Right now, this property is not that important (since drain == just killing raylet), and the purpose of the API is to avoid false negative heartbeat error messages (which doesn't require this property), so I deferred this property for now.

It currently kills the raylet gracefully. It's the implementation detail. Drain != kill raylet, but we are having this behavior for now because that's the least breaking change.

**Please review the schema of the API carefully as we might need to ensure the backward compatibility.**

### Edge cases
Note that this can have "false negative" error message if the raylet has a high load. This should be very very rare. 
- GCS acks, and it stops receiving heartbeat to the raylet.
- GCS still hasn't sent drain requests to raylets (this is the implementation detail). It can be deferred if raylet has high load as well.
- Raylet is killed because of the heartbeat delay (happens in 30 seconds).

As mentioned above, autoscaler can kill the machine before the node receives the drain requests as the API sends this asynchronously. This should have no problem now as the purpose of the API is just to avoid falseful error messages (and this is guaranteed in this scenario), but it can be problematic if we implement proper draining behavior. 

## Kill Raylet implementation
- We send a sigterm to the process. This is the same implementation as ray stop. In this way, we can avoid having multiple shutdown implementation like a core worker. 
- If it is not graceful, we just call `std::_Exit`. 
- This RPC will also be used for chaos testing

## Drain implementation
Currently, Ray already kind of has a concept of drain, with the RPC method "UnregisterNode". This PR is doing
- Refactoring `UnregisterNode` to `DrainNode`.
- Allow to batch draining multiple nodes within a single RPC for the performance.
- gracefully kill raylet before publishing the information to the cluster. Previously, we didn't do this because there was no concept of "draining from central component like GCS". 

## Related issue number

Closes https://github.com/ray-project/ray/issues/19056

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
> RFC part sounds fine to me -- so for now, autoscaler can do a single rpc roundtrip each time it's terminating nodes and we won't get the annoying driver log message.

This is correct.

> Later if we want to make sure a node is drained before terminating, we could terminate nodes a while after draining as suggested by @yiranwang52, following YARN and K8s behavior.

Yep! Note that the only reason why we kill raylets gracefully is because it is the least breaking changes (since we don't know the behavior of Ray cluster well if we don't kill them and just put them as a "drained" state). We can handle this better in the future when we invest more on downscaling.

**Comment by rkooo567** [comment]
> From a cluster's perspective, does the API guarantee the node is drained once it returns? Drained means the cluster doesn't aware of the existence of the node anymore; whether the node is running or not is irrelevant to the cluster.

With the current implementation, "cluster" (GCS + raylet) is still aware of nodes. This is a bit tricky to guarantee because cluster knows the node is drained after the draining information is propagated to all nodes through the pubsub (and we don't have a callback mechanism for pubsub). Also, as I mentioned, if some raylets are having high load issue, the RPC could be extremely slow, which can complicate the API & autoscaler logic (since we need to support timeout in this case. Otherwise, it is possible autoscaler will get reply from GCS extremely slowly). 

Currently, only the GCS (the control plane of the cluster) is guaranteed to know that the node is drained (and everything else will eventually know this). I think this is okay for now because the problem we are facing is that the control plane doesn't know if the node is drained when autoscaler kills the node.

**Comment by rkooo567** [comment]
@DmitriGekhtman addressed the comment.

**Comment by AmeerHajAli** [comment]
@ericl @scv119 , Can one of you please review this?

@rkooo567 , can you maybe give a line or two of code for how this interaction will happen, basically how this translates to python code that autoscaler will write (this will help do it faster).
E.g., in monitor.py today this is how we get the resource usage:
```
        request = gcs_service_pb2.GetAllResourceUsageRequest()
        response = self.gcs_node_resources_stub.GetAllResourceUsage(
            request, timeout=60)
```

**Review by scv119 — APPROVED** [review]
looks good to me! need a bit comment :)

**Review by DmitriGekhtman — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]

File: src/ray/rpc/node_manager/node_manager_client_pool.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ class NodeManagerClientPool {
 
   /// Return an existing NodeManagerWorkerClient if exists, and connect to one if it does
   /// not. The returned pointer is borrowed, and expected to be used briefly.
+  /// The function is guaranteed to return the non-nullptr.
```
Comment (dentiny) [inline src/ray/rpc/node_manager/node_manager_client_pool.h:43]: For these type of in or out values, you could use `absl::Nonnull` for clarify :)

**Pull request merged**